File: app/src/main/java/org/antrack/app/ui/fragments/SmsFragment.java
package org.antrack.app.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.antrack.app.C;
import org.antrack.app.libs.L;
import org.antrack.app.ui.RecyclerViewAnim;
import org.antrack.app.ui.State;
import org.antrack.app.ui.U;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import app.R;

public class SmsFragment extends BaseFragment {
    private final String TAG = "SmsFragment";
    private Context context;

    private ExecutorService executor;

    private List<Sms> smses;
    private SmsAdapter smsAdapter;
    private RecyclerViewAnim recyclerView;

    String smsDir;
    String smsCmd;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Otherwise GetActivity() return null after orientation change
        setRetainInstance(true);

        if (!Mod.check(Mod.DUMPSMS)) {
            showNoModule(Mod.DUMPSMS);
            return null;
        }

        if (!State.features.phone) {
            showNoPhone();
            return null;
        }

        smsDir = Mod.getFile(Mod.DUMPSMS);
        smsCmd = Mod.getCommand(Mod.DUMPSMS);

        context = getActivity().getApplicationContext();

        View view = inflater.inflate(R.layout.fragment_cardview, container, false);

        recyclerView = (RecyclerViewAnim) view.findViewById(R.id.fragment_cardview_list);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context);
        recyclerView.setLayoutManager(linearLayoutManager);

        smses = new ArrayList<>();
        smsAdapter = new SmsAdapter(smses);
        recyclerView.setAdapter(smsAdapter);

        executor = Executors.newFixedThreadPool(1);

        onFileUpdate();

        U.runCommandAsync(smsCmd);

        if (!State.device.isMain()) {
            // FIXME
            U.getFileAsync(smsDir + "inbox");
            U.getFileAsync(smsDir + "sent");
        }

        return view;
    }

    @Override
    public String getWatchFile() {
        return smsDir;
    }

    @Override
    public void onFileUpdate() {
        executor.submit(new Runnable() {
            @Override
            public void run() {
                waitCardsDrawn(recyclerView);

                smses = new ArrayList<>();

                if (!readFile()) {
                    showNoData();
                    return;
                }

                if (getActivity() == null) return;
                getActivity().runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        smsAdapter.update(smses);
                        smsAdapter.notifyDataSetChanged();
                        hideAllMessages();
                    }
                });
            }
        });
    }

    public class SmsComaparator implements Comparator<Sms> {
        @Override
        public int compare(Sms o1, Sms o2) {
            Date date1, date2;

            SimpleDateFormat format = new SimpleDateFormat(C.DEFAULT_TIME_FORMAT);
            try {
                date1 = format.parse(o1.date);
                date2 = format.parse(o2.date);
            } catch (Exception e) {
                L.e(TAG, "DateComparator exception: " + e.toString());
                return -1;
            }
            return date1.compareTo(date2);
        }
    }

    private boolean readFile() {
        // FIXME
        String pathIn = U.getLocalPath(smsDir + "inbox");
        String pathOut = U.getLocalPath(smsDir + "sent");

        if (!new File(pathIn).exists() && !new File(pathOut).exists()) {
            return false;
        }

        ReadBox(pathIn, true);
        ReadBox(pathOut, false);

        if (smses.isEmpty()) {
            return false;
        }

        // FIXME DateComparator exception: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        Collections.sort(smses, new SmsComaparator());
        Collections.reverse(smses);

        return true;
    }

    private boolean ReadBox(String path, final boolean inbox) {
        try {
            BufferedReader reader = new BufferedReader(new FileReader(path));
            String line;
            Sms sms = new Sms();
            while ((line = reader.readLine()) != null) {
                String[] pair = line.split(":");
                switch(pair[0]) {
                    case "From":
                        sms.from = line.replaceAll(pair[0] + ":", "").trim();
                        break;
                    case "Date":
                        sms.date = line.replaceAll(pair[0] + ":", "").trim();
                        break;
                    case "Body":
                        sms.body = line.replaceAll(pair[0] + ":", "").trim();
                        break;
                    default:
                        if (inbox) {
                            sms.direction = "Incoming";
                        } else {
                            sms.direction = "Outgoing";
                        }
                        smses.add(sms);
                        sms = new Sms();
                }
            }
        } catch (IOException e) {
            L.e(TAG, "Can't read sms file: " + e);
            return false;
        }
        return true;
    }
}

File: app/src/main/java/org/antrack/app/ui/fragments/SmsFragment.kt
package org.antrack.app.ui.fragments

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import app.R
import org.antrack.app.C
import org.antrack.app.libs.L
import org.antrack.app.ui.RecyclerViewAnim
import org.antrack.app.ui.State
import org.antrack.app.ui.U
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.collections.ArrayList

class SmsFragment : BaseFragment() {
    private val TAG = "SmsFragment"

    override val module = Mod.DUMPSMS

    lateinit private var executor: ExecutorService
    lateinit private var smsAdapter: SmsAdapter
    lateinit private var recyclerView: RecyclerViewAnim

    private var smses = ArrayList<Sms>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Otherwise GetActivity() return null after orientation change
        retainInstance = true

        checkModule() || return null
        checkPhone() || return null

        val context = activity.applicationContext

        val view = inflater.inflate(R.layout.fragment_cardview, container, false)

        recyclerView = view.findViewById(R.id.fragment_cardview_list) as RecyclerViewAnim
        val linearLayoutManager = LinearLayoutManager(context)
        recyclerView.layoutManager = linearLayoutManager

        smses = ArrayList<Sms>()
        smsAdapter = SmsAdapter(smses)
        recyclerView.adapter = smsAdapter

        executor = Executors.newFixedThreadPool(1)

        onFileUpdate()

        command?.let {
            U.runCommandAsync(it)
        }

        if (!State.device.isMain) {
            watchFile?.let {
                U.getFileAsync(it + "inbox")
                U.getFileAsync(it + "sent")
            }
        }

        return view
    }

    override fun onFileUpdate() {
        executor.submit(Runnable {
            waitCardsDrawn(recyclerView)

            smses = ArrayList<Sms>()

            if (!readFile() || smses.isEmpty()) {
                showNoDataOrLoading()
                return@Runnable
            }

            if (activity == null) return@Runnable
            activity.runOnUiThread {
                smsAdapter.update(smses)
                smsAdapter.notifyDataSetChanged()
                hideAllMessages()
            }
        })
    }

    inner class SmsComparator : Comparator<Sms> {
        override fun compare(o1: Sms, o2: Sms): Int {
            val date1: Date
            val date2: Date

            val format = SimpleDateFormat(C.DEFAULT_TIME_FORMAT, Locale.US)
            try {
                date1 = format.parse(o1.date)
                date2 = format.parse(o2.date)
            } catch (e: Exception) {
                e.printStackTrace()
                return -1
            }

            return date1.compareTo(date2)
        }
    }

    private fun readFile(): Boolean {
        // FIXME
        val pathIn = U.getLocalPath(watchFile!! + "inbox")
        val pathOut = U.getLocalPath(watchFile!! + "sent")

        if (!File(pathIn).exists() && !File(pathOut).exists()) {
            return false
        }

        ReadBox(pathIn, true)
        ReadBox(pathOut, false)

        if (smses.isEmpty()) {
            return false
        }

        // FIXME DateComparator exception: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        Collections.sort(smses, SmsComparator())
        Collections.reverse(smses)

        return true
    }

    private fun ReadBox(path: String, inbox: Boolean): Boolean {
        try {
            val reader = BufferedReader(FileReader(path))
            var sms = Sms()

            reader.readLines().forEach { line ->
                val pair = line.split(":".toRegex())
                when (pair[0]) {
                    "From", "To" -> sms.from = line.replace((pair[0] + ":").toRegex(), "").trim { it <= ' ' }
                    "Date" -> sms.date = line.replace((pair[0] + ":").toRegex(), "").trim { it <= ' ' }
                    "Body" -> sms.body = line.replace((pair[0] + ":").toRegex(), "").trim { it <= ' ' }
                    else -> {
                        if (inbox) {
                            sms.direction = "Incoming"
                        } else {
                            sms.direction = "Outgoing"
                        }
                        if (sms.from != null && sms.date != null && sms.body != null) {
                            smses.add(sms)
                        }
                        sms = Sms()
                    }
                }
            }
        } catch (e: IOException) {
            L.e(TAG, "Can't read sms file: " + e)
            return false
        }

        return true
    }
}
